hack/generated/controllers/suite_test.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package controllers_test
+
+import (
+	"context"
+	"log"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	resources "github.com/Azure/k8s-infra/hack/generated/apis/microsoft.resources/v20200601"
+	"github.com/Azure/k8s-infra/hack/generated/controllers"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/testcommon"
+)
+
+const (
+	TestRegion    = "westus"
+	TestNamespace = "k8s-infra-test-ns"
+)
+
+type ControllerTestContext struct {
+	*testcommon.TestContext
+	SharedResourceGroup *resources.ResourceGroup
+}
+
+func setup() (ControllerTestContext, error) {
+	ctx := context.Background()
+	log.Println("Running test setup")
+
+	SetDefaultEventuallyTimeout(2 * time.Minute)
+	SetDefaultEventuallyPollingInterval(5 * time.Second)
+
+	testContext, err := testcommon.NewTestContext(
+		TestRegion,
+		TestNamespace,
+		controllers.ResourceStateAnnotation,
+		controllers.ResourceErrorAnnotation)
+
+	if err != nil {
+		return ControllerTestContext{}, err
+	}
+
+	err = testContext.CreateTestNamespace()
+	if err != nil {
+		return ControllerTestContext{}, err
+	}
+
+	// Create a shared resource group, for tests to use
+	sharedResourceGroup := testContext.NewTestResourceGroup()
+	err = testContext.KubeClient.Create(ctx, sharedResourceGroup)
+	if err != nil {
+		return ControllerTestContext{}, errors.Wrapf(err, "creating shared resource group")
+	}
+
+	// TODO: Should use AzureName here once it's always set
+	log.Printf("Created shared resource group %s\n", sharedResourceGroup.Name)
+
+	// It should be created in Kubernetes
+	// TODO: Generic polling helper needed here
+	// g.Eventually(testContext.Ensure.Created(ctx, rg)).Should(BeTrue())
+
+	log.Println("Done with test setup")
+
+	return ControllerTestContext{testContext, sharedResourceGroup}, nil
+}
+
+func teardown(testContext ControllerTestContext) error {
+	log.Println("Started common controller test teardown")
+
+	ctx := context.Background()
+
+	// List all of the resource groups
+	rgList := &resources.ResourceGroupList{}
+	err := testContext.KubeClient.List(ctx, rgList)
+	if err != nil {
+		return errors.Wrap(err, "listing resource groups")
+	}
+
+	// Delete any leaked resource groups
```

Comment:
How do we know that they leaked from us?

Would be really bad form to delete any resource groups set up that aren't to do with our operator.

Reply:
By definition the list of resources discovered in this cleanup method is from Kubernetes, so currently the assumption is that we own the cluster and it's being solely used for testing.

Since we have a specific test namespace I'll limit this to querying just that, so that it'll play nicer in a cluster being used for something else as well.